3.8/graphs-custom-pregel.md

```diff
@@ -0,0 +1,1285 @@
+ = 12
+> ["apply", "min", ["quote", 1, 2, 3]]
+ = 1
+> ["map", "int-to-string", ["quote", 1, 2, 3, 4]]
+ = ["1", "2", "3", "4"]
+```
+
+#### Variables
+
+```js
+["var-ref", name]
+["bind-ref", name]
+```
+
+`var-ref` evaluates to the current value of the variable with name `name`.
+It is an error to reference a variable that is not defined in the current
+context. `bind-ref` is an alias of `var-ref`.
+
+#### Debug operators
+
+```js
+["print", values...]
+["error", msg...]
+["assert", cond, msg...]
+```
+
+`print` print in an context dependent way the string representation of its
+arguments joined by spaces. Strings represent themselves, numbers are converted
+to decimal representation, booleans are represented as `true` or `false`.
+Dicts and lists are converted to JSON.
+
+`error` creates an error an aborts execution immediately. Errors are reported
+in a context dependent way. The error message is constructed from the remaining
+parameter like `print`, except that it is not printed but associated with the
+error. This like a panic or an uncaught exception.
+
+`assert` checks if cond is considered true if it an error with the remaining
+parameters as message is raised. It is equivalent to
+`["if", [cond, ["error", msg...]]]`.
+
+### Foreign calls in _Vertex Computation_ context
+
+The following functions are only available when running as a vertex computation
+(i.e. as a `initProgram`, `updateProgram`, ...). `this` usually refers to the
+vertex we are attached to.
+
+#### Local Accumulators
+
+```js
+["accum-ref", name]
+["accum-set!", name, value]
+["accum-clear!", name]
+```
+
+- `accum-ref` evaluates to the current value of the accumulator `name`.
+- `accum-set!` sets the current value of the accumulator `name` to `value`.
+- `accum-clear!` resets the current value of the accumulator `name` to a
+  well-known one. Currently numeric limits for
+- `max` and `min` accumulators, 0 for sums, false for or, true for and, and
+  empty for lists and VelocyPack.
+
+#### Global Accumulators
+
+```
+["global-accum-ref", name]
+["send-to-global-accum", name, value]
+```
+
+- `global-accum-ref` evaluates the global accumulator `name`.
+- `send-to-global-accum` send `value` to the global accumulator `name`.
+
+Also see the remarks about [update visibility](#vertex-accumulators).
+
+#### Message Passing
+
+```js
+["send-to-accum", name, to-pregel-id, value]
+["send-to-all-neighbors", name, value]
+```
+
+- `send-to-accum` send the value `value` to the accumulator `name` at vertex
+  with pregel-id `to-pregel-vertex`.
+- `send-to-all-neighbors` send the value `value` to the accumulator `name` in
+  all neighbors reachable by an edge. Note that if there are multiple edges
+  from us to the neighbor, the value is sent multiple times.
+
+#### This Vertex
+
+```js
+["this-doc"]
+["this-vertex-id"]
+["this-unique-id"]
+["this-pregel-id"]
+["this-outdegree"]
+["this-outbound-edges"]
+```
+
+- `this-doc` returns the document slice stored in vertex data.
```

Comment:
I think we need to elaborate on the document slice we have in our hands here now.

Does it contain the full document?
Does it contain only what we have defined by data access?
What is the format?

An example of readValue on Document Data and one on Edge Data would be very helpful here.